src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ObjectDepthTooLarge);
+
+            _consumed++;
+            _lineBytePosition++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+                _stackFreeContainer = (_stackFreeContainer << 1) | 1;
+            else
+                _stack.Push(JsonTokenType.StartObject);
+
+            _tokenType = JsonTokenType.StartObject;
+            _inObject = true;
+        }
+
+        private void EndObject()
+        {
+            if (!_inObject || _currentDepth <= 0)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ObjectEndWithinArray);
+
+            _consumed++;
+            _lineBytePosition++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+            {
+                _stackFreeContainer >>= 1;
+                _inObject = (_stackFreeContainer & 1) != 0;
+            }
+            else
+            {
+                _inObject = _stack.Pop() != JsonTokenType.StartArray;
+            }
+
+            _currentDepth--;
+            _tokenType = JsonTokenType.EndObject;
+        }
+
+        private void StartArray()
+        {
+            _currentDepth++;
+            if (_currentDepth > _maxDepth)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ArrayDepthTooLarge);
+
+            _consumed++;
+            _lineBytePosition++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+                _stackFreeContainer = _stackFreeContainer << 1;
+            else
+                _stack.Push(JsonTokenType.StartArray);
+
+            _tokenType = JsonTokenType.StartArray;
+            _inObject = false;
+        }
+
+        private void EndArray()
+        {
+            if (_inObject || _currentDepth <= 0)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ArrayEndWithinObject);
+
+            _consumed++;
+            _lineBytePosition++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+            {
+                _stackFreeContainer >>= 1;
+                _inObject = (_stackFreeContainer & 1) != 0;
+            }
+            else
+            {
+                _inObject = _stack.Pop() != JsonTokenType.StartArray;
+            }
+
+            _currentDepth--;
+            _tokenType = JsonTokenType.EndArray;
+        }
+
+        private bool ReadSingleSegment()
+        {
+            bool retVal = false;
+            IsValueMultiSegment = false;
+
+            if (!HasMoreData())
+                goto Done;
+
+            byte first = _buffer[_consumed];
+
+            if (first <= JsonConstants.Space)
+            {
+                SkipWhiteSpace();
+                if (!HasMoreData())
+                    goto Done;
+                first = _buffer[_consumed];
+            }
+
+            if (_tokenType == JsonTokenType.None)
+                goto ReadFirstToken;
```

Comment:
Does this extra goto really help over just having the if body have the two lines of the trailing label?

Reply:
Yes. ~2% regression otherwise. Not large enough to incur the complexity on its own, but this method already crossed that threshold, imo.